File: js/logbrew-node/examples/first-useful-telemetry.mjs
import { createServer } from "node:http";
import { once } from "node:events";
import {
  createNetworkMilestoneAttributes,
  createProductActionAttributes,
  createTraceparentHeaders,
  parseTraceparent,
  RecordingTransport
} from "@logbrew/sdk";
import {
  createLogBrewNodeClient,
  withLogBrewHttpHandler
} from "@logbrew/node";

const traceId = "4bf92f3577b34da6a3ce929d0e0e4736";
const inboundHeaders = createTraceparentHeaders({
  traceId,
  spanId: "00f067aa0ba902b7"
});
const routeTemplate = "/checkout/:cartId";
const transport = RecordingTransport.alwaysAccept();
const startupClient = createLogBrewNodeClient({
  serverApiKey: "LOGBREW_SERVER_API_KEY",
  sdkName: "checkout-api",
  sdkVersion: "1.4.0"
});

startupClient.release("evt_release_checkout_api", "2026-06-15T08:00:00Z", {
  version: "1.4.0",
  commit: "abc123def456",
  metadata: {
    service: "checkout-api"
  }
});
startupClient.environment("evt_environment_checkout_api", "2026-06-15T08:00:01Z", {
  name: "production",
  region: "us-east-1"
});
await startupClient.flush(transport);

let monotonicMs = 0;
const server = createServer(withLogBrewHttpHandler((req, res, logbrew) => {
  const requestTraceId = traceIdFromRequest(req);
  logbrew.client.log("evt_log_checkout_received", "2026-06-15T08:00:02Z", {
    message: "checkout request accepted",
    level: "info",
    logger: "checkout-api",
    metadata: {
      method: "POST",
      routeTemplate,
      traceId: requestTraceId
    }
  });
  logbrew.client.action(
    "evt_action_checkout_started",
    "2026-06-15T08:00:03Z",
    createProductActionAttributes({
      name: "checkout started",
      status: "running",
      sessionId: "sess_checkout_123",
      traceId: requestTraceId,
      routeTemplate,
      funnel: "checkout",
      step: "payment"
    })
  );
  logbrew.client.action(
    "evt_network_payment_authorized",
    "2026-06-15T08:00:04Z",
    createNetworkMilestoneAttributes({
      routeTemplate: "/payments/:paymentId",
      method: "POST",
      status: "success",
      statusCode: 202,
      durationMs: 43,
      sessionId: "sess_checkout_123",
      traceId: requestTraceId
    })
  );
  logbrew.client.metric("evt_metric_checkout_duration", "2026-06-15T08:00:05Z", {
    name: "checkout.duration",
    kind: "histogram",
    value: 128,
    unit: "ms",
    temporality: "delta",
    metadata: {
      routeTemplate,
      traceId: requestTraceId
    }
  });
  res.statusCode = 202;
  res.end("accepted");
}, {
  idFactory: () => "evt_span_checkout_request",
  now: () => "2026-06-15T08:00:06Z",
  nowMs: () => {
    monotonicMs += 17;
    return monotonicMs;
  },
  sdkName: "checkout-api",
  sdkVersion: "1.4.0",
  serverApiKey: "LOGBREW_SERVER_API_KEY",
  spanIdFactory: () => "b7ad6b7169203331",
  transport
}));

server.listen(0);
await once(server, "listening");
const port = server.address().port;
const response = await fetch(`http://127.0.0.1:${port}/checkout/123?coupon=summer`, {
  headers: inboundHeaders,
  method: "POST"
});
if (response.status !== 202) {
  throw new Error(`unexpected checkout status: ${response.status}`);
}
await waitFor(() => transport.sentBodies.length === 2);
await closeServer(server);

const batches = transport.sentBodies.map((body) => JSON.parse(body));
const payload = {
  sdk: batches[0].sdk,
  events: batches.flatMap((batch) => batch.events)
};
const eventIds = payload.events.map((event) => event.id);
if (eventIds.length !== 7) {
  throw new Error(`unexpected event count: ${eventIds.join(",")}`);
}
const requestSpan = payload.events.find((event) => event.id === "evt_span_checkout_request");
if (!requestSpan || requestSpan.type !== "span") {
  throw new Error(`missing request span: ${JSON.stringify(payload)}`);
}
if (requestSpan.attributes.traceId !== traceId || requestSpan.attributes.parentSpanId !== "00f067aa0ba902b7") {
  throw new Error(`unexpected trace context: ${JSON.stringify(requestSpan)}`);
}
if (requestSpan.attributes.metadata.path !== "/checkout/123") {
  throw new Error(`request path should omit query text: ${JSON.stringify(requestSpan)}`);
}
const networkEvent = payload.events.find((event) => event.id === "evt_network_payment_authorized");
if (networkEvent?.attributes.metadata.routeTemplate !== "/payments/:paymentId") {
  throw new Error(`missing network milestone route template: ${JSON.stringify(networkEvent)}`);
}

console.log(JSON.stringify(payload, null, 2));
console.error(JSON.stringify({
  ok: true,
  events: eventIds.length,
  requestSpan: requestSpan.attributes.name,
  networkMilestone: networkEvent.attributes.name,
  traceId: requestSpan.attributes.traceId
}));

function traceIdFromRequest(req) {
  const value = req.headers.traceparent;
  if (typeof value !== "string") {
    return undefined;
  }
  try {
    return parseTraceparent(value).traceId;
  } catch {
    return undefined;
  }
}

async function waitFor(predicate) {
  for (let attempt = 0; attempt < 20; attempt += 1) {
    if (predicate()) {
      return;
    }
    await new Promise((resolve) => {
      setTimeout(resolve, 10);
    });
  }
  throw new Error("timed out waiting for Node.js telemetry");
}

async function closeServer(serverToClose) {
  await new Promise((resolve, reject) => {
    serverToClose.close((error) => {
      if (error) {
        reject(error);
        return;
      }
      resolve();
    });
  });
}

File: python/logbrew_py/src/logbrew_sdk/examples/first_useful_telemetry.py
"""Packaged first-useful telemetry example for installed Python users."""

from __future__ import annotations

import json
import logging
import sys

from logbrew_sdk import (
    LogBrewClient,
    LogBrewLoggingHandler,
    RecordingTransport,
    create_network_milestone_attributes,
    create_product_action_attributes,
    create_traceparent_headers,
    parse_traceparent,
    span_attributes_from_traceparent,
)


def main() -> int:
    trace_id = "4bf92f3577b34da6a3ce929d0e0e4736"
    inbound_headers = create_traceparent_headers(
        trace_id=trace_id,
        span_id="00f067aa0ba902b7",
    )
    route_template = "/checkout/:cart_id"
    client = LogBrewClient.create(
        api_key="LOGBREW_API_KEY",
        sdk_name="checkout-api",
        sdk_version="1.4.0",
    )

    client.release(
        "evt_release_checkout_api",
        "2026-06-15T08:00:00Z",
        {
            "version": "1.4.0",
            "commit": "abc123def456",
            "metadata": {"service": "checkout-api"},
        },
    )
    client.environment(
        "evt_environment_checkout_api",
        "2026-06-15T08:00:01Z",
        {"name": "production", "region": "us-east-1"},
    )

    logger = logging.getLogger("checkout-api")
    logger.propagate = False
    logger.setLevel(logging.INFO)
    logger.addHandler(
        LogBrewLoggingHandler(
            client,
            metadata={"service": "checkout-api"},
        )
    )

    handle_checkout_request(
        client,
        logger,
        traceparent=inbound_headers["traceparent"],
        route_template=route_template,
    )

    payload = json.loads(client.preview_json())
    assert_first_useful_payload(payload, trace_id=trace_id)

    print(json.dumps(payload, indent=2))
    response = client.shutdown(RecordingTransport.always_accept())
    print(
        json.dumps(
            {
                "ok": True,
                "status": response.status_code,
                "attempts": response.attempts,
                "events": len(payload["events"]),
                "requestSpan": "evt_span_checkout_request",
                "traceId": trace_id,
            },
            sort_keys=True,
        ),
        file=sys.stderr,
    )
    return 0


def handle_checkout_request(
    client: LogBrewClient,
    logger: logging.Logger,
    *,
    traceparent: str,
    route_template: str,
) -> None:
    trace_context = parse_traceparent(traceparent)

    logger.info(
        "checkout request accepted",
        extra={
            "method": "POST",
            "routeTemplate": route_template,
            "traceId": trace_context.trace_id,
        },
    )
    client.action(
        "evt_action_checkout_started",
        "2026-06-15T08:00:03Z",
        create_product_action_attributes(
            {
                "name": "checkout started",
                "status": "running",
                "sessionId": "sess_checkout_123",
                "traceId": trace_context.trace_id,
                "routeTemplate": f"{route_template}?coupon=private#payment",
                "funnel": "checkout",
                "step": "payment",
                "metadata": {"service": "checkout-api", "payload": {"card": "redacted"}},
            }
        ),
    )
    client.action(
        "evt_network_payment_authorized",
        "2026-06-15T08:00:04Z",
        create_network_milestone_attributes(
            {
                "routeTemplate": "https://api.example.test/payments/:payment_id?card=private#receipt",
                "method": "post",
                "statusCode": 202,
                "durationMs": 43,
                "sessionId": "sess_checkout_123",
                "traceId": trace_context.trace_id,
                "metadata": {"service": "checkout-api", "headers": {"authorization": "redacted"}},
            }
        ),
    )
    client.metric(
        "evt_metric_checkout_duration",
        "2026-06-15T08:00:05Z",
        {
            "name": "checkout.duration",
            "kind": "histogram",
            "value": 128,
            "unit": "ms",
            "temporality": "delta",
            "metadata": {
                "routeTemplate": route_template,
                "traceId": trace_context.trace_id,
            },
        },
    )
    client.span(
        "evt_span_checkout_request",
        "2026-06-15T08:00:06Z",
        span_attributes_from_traceparent(
            traceparent,
            name="POST /checkout/:cart_id",
            span_id="b7ad6b7169203331",
            status="ok",
            duration_ms=17,
            metadata={
                "method": "POST",
                "routeTemplate": route_template,
                "service": "checkout-api",
            },
        ),
    )


def assert_first_useful_payload(payload: dict[str, object], *, trace_id: str) -> None:
    events = payload.get("events")
    if not isinstance(events, list):
        raise RuntimeError(f"missing events: {payload}")
    event_ids = [event.get("id") for event in events if isinstance(event, dict)]
    if len(event_ids) != 7 or event_ids[:2] != [
        "evt_release_checkout_api",
        "evt_environment_checkout_api",
    ] or event_ids[3:] != [
        "evt_action_checkout_started",
        "evt_network_payment_authorized",
        "evt_metric_checkout_duration",
        "evt_span_checkout_request",
    ]:
        raise RuntimeError(f"unexpected event order: {event_ids}")
    log_event = events[2]
    if (
        not isinstance(log_event, dict)
        or log_event.get("type") != "log"
        or not str(log_event.get("id", "")).startswith("evt_log_checkout_api_")
    ):
        raise RuntimeError(f"unexpected log event: {log_event}")

    output = json.dumps(payload)
    for forbidden in ("coupon=private", "card=private", "authorization", "redacted"):
        if forbidden in output:
            raise RuntimeError(f"first useful telemetry leaked private data: {forbidden}")

    request_span = find_event(events, "evt_span_checkout_request")
    request_attributes = require_attributes(request_span)
    if request_attributes["traceId"] != trace_id:
        raise RuntimeError(f"unexpected request span trace: {request_span}")
    if request_attributes["parentSpanId"] != "00f067aa0ba902b7":
        raise RuntimeError(f"unexpected request span parent: {request_span}")

    network_event = find_event(events, "evt_network_payment_authorized")
    network_attributes = require_attributes(network_event)
    network_metadata = network_attributes.get("metadata")
    if not isinstance(network_metadata, dict) or network_metadata.get("routeTemplate") != "/payments/:payment_id":
        raise RuntimeError(f"unexpected network route template: {network_event}")


def find_event(events: list[object], event_id: str) -> dict[str, object]:
    for event in events:
        if isinstance(event, dict) and event.get("id") == event_id:
            return event
    raise RuntimeError(f"missing event: {event_id}")


def require_attributes(event: dict[str, object]) -> dict[str, object]:
    attributes = event.get("attributes")
    if not isinstance(attributes, dict):
        raise RuntimeError(f"missing attributes: {event}")
    return attributes


if __name__ == "__main__":
    raise SystemExit(main())
